hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStream.java

```diff
@@ -285,14 +311,22 @@ private synchronized void writeCurrentBufferToService() throws IOException {
 
     final byte[] bytes = buffer;
     final int bytesLength = bufferIndex;
-    buffer = byteBufferPool.getBuffer(false, bufferSize).array();
     bufferIndex = 0;
     final long offset = position;
     position += bytesLength;
 
-    if (threadExecutor.getQueue().size() >= maxConcurrentRequestCount * 2) {
+    if (Runtime.getRuntime().freeMemory() < MIN_MEMORY_THRESHOLD
+        && buffersToBeReturned.get() > 0) {
+      LOG.debug("Low memory");
```

Comment:
what if` Runtime.getRuntime().freeMemory() < MIN_MEMORY_THRESHOLD` but `buffersToBeReturned.get()` is `0`?
Can we make the LOG more meaningful like  "Low memory, less than 100 MB is available, the existing read tasks consumes 2048 MB, waiting for completion."